### PR TITLE
fix: track metadata, score, prompt, and create group #16

### LINKED ISSUE #16: track metadata, score, prompt, and create group
We have exposed multiple functions in our python sdk

- track metadata
- track score
- track prompt
- create group

We need a replica of such functions in our javascript sdk.

### PR DESCRIPTION
**Pull request by abubakarsohail**
fix #16

This PR adds support for tracking features in PromptLayer. This includes tracking metadata, score, prompt and group.

### CODE REVIEW

File: src/track/index.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        "Please provide a dictionary of metadata with key value pair of strings."
+      );
+    }
+  }
+  return promptLayerTrackMetadata(body);
+};
+
+const score = (body: track.Score): Promise<boolean> => {
+  if (typeof body.score !== "number") {
+    throw new Error("Please provide a int score.");
```
Comment (jzone3) [inline src/track/index.ts:25]: "Score must be a number"

File: src/track/index.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  }
+  if (body.score < 0 || body.score > 100) {
+    throw new Error("Please provide a score between 0 and 100.");
```
Comment (jzone3) [inline src/track/index.ts:28]: "Score must be a number between 0 and 100."

File: src/track/index.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  }
+  return promptLayerTrackScore(body);
+};
+
+const prompt = (body: track.Prompt): Promise<boolean> => {
+  if (!(body.prompt_input_variables instanceof Object)) {
+    throw new Error("Please provide a dictionary of input variables.");
```
Comment (jzone3) [inline src/track/index.ts:35]: "Prompt template input variable dictionary not provided."

File: src/types/track.ts [hunk 4]
```diff
@@ -14,3 +14,26 @@ export interface TrackRequest {
...
+  request_id: number;
+  metadata: Record<string, string>;
+}
+
+export interface Score {
+  request_id: number;
+  score: number;
+}
+
+export interface Prompt {
```
Comment (jzone3) [inline src/types/track.ts:28]: PromptTemplate

File: src/utils/index.ts [hunk 5]
```diff
@@ -36,10 +36,9 @@ const promptLayerApiRequest = async (body: TrackRequest) => {
     });
     const data = await response.json();
     if (response.status !== 200) {
-      console.warn(
-        `WARNING: While logging your request PromptLayer had the following error: ${JSON.stringify(
-          data
-        )}`
+      warnOnBadResponse(
+        data,
+        "WARNING: While logging your request PromptLayer had the following error"
```
Comment (jzone3) [inline src/utils/index.ts:41]: "WARNING: While logging your request, PromptLayer experienced the following error:"

File: src/utils/index.ts [hunk 6]
```diff
@@ -53,6 +52,158 @@ const promptLayerApiRequest = async (body: TrackRequest) => {
...
+          ...body,
+          api_key: getApiKey(),
+        }),
+      }
+    );
+    const data = await response.json();
+    if (response.status !== 200) {
+      warnOnBadResponse(
+        data,
+        "WARNING: While tracking your metadata PromptLayer had the following error"
```
Comment (jzone3) [inline src/utils/index.ts:76]: "WARNING: While logging metadata to your request, PromptLayer experienced the following error"

File: src/utils/index.ts [hunk 7]
```diff
@@ -53,6 +52,158 @@ const promptLayerApiRequest = async (body: TrackRequest) => {
...
+      );
+      return false;
+    }
+  } catch (e) {
+    console.warn(
+      `WARNING: While tracking your metadata PromptLayer had the following error: ${e}`
```
Comment (jzone3) [inline src/utils/index.ts:82]: "WARNING: While logging metadata to your request, PromptLayer experienced the following error"

File: src/utils/index.ts [hunk 8]
```diff
@@ -53,6 +52,158 @@ const promptLayerApiRequest = async (body: TrackRequest) => {
...
+      body: JSON.stringify({
+        ...body,
+        api_key: getApiKey(),
+      }),
+    });
+    const data = await response.json();
+    if (response.status !== 200) {
+      warnOnBadResponse(
+        data,
+        "WARNING: While tracking your score PromptLayer had the following error"
```
Comment (jzone3) [inline src/utils/index.ts:105]: "WARNING: While scoring your request, PromptLayer experienced the following error"

File: src/utils/index.ts [hunk 9]
```diff
@@ -53,6 +52,158 @@ const promptLayerApiRequest = async (body: TrackRequest) => {
...
+      );
+      return false;
+    }
+  } catch (e) {
+    console.warn(
+      `WARNING: While tracking your score PromptLayer had the following error: ${e}`
```
Comment (jzone3) [inline src/utils/index.ts:111]: "WARNING: While scoring your request, PromptLayer experienced the following error"

File: src/utils/index.ts [hunk 10]
```diff
@@ -53,6 +52,158 @@ const promptLayerApiRequest = async (body: TrackRequest) => {
...
+          ...body,
+          api_key: getApiKey(),
+        }),
+      }
+    );
+    const data = await response.json();
+    if (response.status !== 200) {
+      warnOnBadResponse(
+        data,
+        "WARNING: While tracking your prompt PromptLayer had the following error"
```
Comment (jzone3) [inline src/utils/index.ts:137]: "WARNING: While associating your request with a prompt template, PromptLayer experienced the following error"

File: src/utils/index.ts [hunk 11]
```diff
@@ -53,6 +52,158 @@ const promptLayerApiRequest = async (body: TrackRequest) => {
...
+      );
+      return false;
+    }
+  } catch (e) {
+    console.warn(
+      `WARNING: While tracking your prompt PromptLayer had the following error: ${e}`
```
Comment (jzone3) [inline src/utils/index.ts:143]: "WARNING: While associating your request with a prompt template, PromptLayer experienced the following error"

File: src/utils/index.ts [hunk 12]
```diff
@@ -53,6 +52,158 @@ const promptLayerApiRequest = async (body: TrackRequest) => {
...
+      body: JSON.stringify({
+        ...body,
+        api_key: getApiKey(),
+      }),
+    });
+    const data = await response.json();
+    if (response.status !== 200) {
+      warnOnBadResponse(
+        data,
+        "WARNING: While tracking your group PromptLayer had the following error"
```
Comment (jzone3) [inline src/utils/index.ts:166]: "WARNING: While associating your request with a group, PromptLayer experienced the following error"

File: src/utils/index.ts [hunk 13]
```diff
@@ -53,6 +52,158 @@ const promptLayerApiRequest = async (body: TrackRequest) => {
...
+      );
+      return false;
+    }
+  } catch (e) {
+    console.warn(
+      `WARNING: While tracking your group PromptLayer had the following error: ${e}`
```
Comment (jzone3) [inline src/utils/index.ts:172]: "WARNING: While associating your request with a group, PromptLayer experienced the following error"

File: src/utils/index.ts [hunk 14]
```diff
@@ -53,6 +52,158 @@ const promptLayerApiRequest = async (body: TrackRequest) => {
...
+      },
+      body: JSON.stringify({
+        api_key: getApiKey(),
+      }),
+    });
+    const data = await response.json();
+    if (response.status !== 200) {
+      warnOnBadResponse(
+        data,
+        "WARNING: While creating your group PromptLayer had the following error"
```
Comment (jzone3) [inline src/utils/index.ts:194]: "a group,"

File: src/utils/index.ts [hunk 15]
```diff
@@ -53,6 +52,158 @@ const promptLayerApiRequest = async (body: TrackRequest) => {
...
+      );
+      return false;
+    }
+    return data.id;
+  } catch (e) {
+    console.warn(
+      `WARNING: While creating your group PromptLayer had the following error: ${e}`
```
Comment (jzone3) [inline src/utils/index.ts:201]: "a group,"

**Review by jzone3 — DISMISSED** [review]
Just fix the error messages then good to go!

File: src/track/index.ts (hunk 1 above)
Comment (abubakarsohail) [reply src/track/index.ts:25]: Updated.

File: src/track/index.ts (hunk 2 above)
Comment (abubakarsohail) [reply src/track/index.ts:28]: Updated.

File: src/track/index.ts (hunk 3 above)
Comment (abubakarsohail) [reply src/track/index.ts:35]: Updated

File: src/utils/index.ts (hunk 5 above)
Comment (abubakarsohail) [reply src/utils/index.ts:41]: Updated.

File: src/utils/index.ts (hunk 6 above)
Comment (abubakarsohail) [reply src/utils/index.ts:76]: Updated.

File: src/utils/index.ts (hunk 7 above)
Comment (abubakarsohail) [reply src/utils/index.ts:82]: Updated.

**Review by jzone3 — DISMISSED** [review]

**Pull request merged**